### PR TITLE
Simplify Hive metastore creation and use

### PR DESCRIPTION
**Pull request by dain**
## Description
* Delay creation of actual `ConnectorMetadata` actually until used
* Provide `ConnectorSession` during creation of `ConnectorMetadata`
* Delay `HiveMetastore` creation until `ConnectorIdentity` is available from `ConnectorSession`
* Capture `ConnectorIdentity` curing metastore creation, and use this instead of passing to every method
* Redesign recording and caching metastore wrapper to make this possible

## General information

Is this change a fix, improvement, new feature, refactoring, or other?
refactoring

Is this a change to the core query engine, a connector, client library, or the
SPI interfaces (be specific)?
engine and spi

How would you describe this change to a non-technical end user or system
administrator?
N/A

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```
# SPI
* Add `ConnectorSession` to `Connector` `getMetadata` method. The former signature is deprecated and should be updated. ({issue}`9482`)
```

### CODE REVIEW

**Comment by electrum** [comment]
Typo "Guild" in commit description

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/SharedHiveMetastoreCache.java [hunk 1]
```diff
@@ -212,6 +217,201 @@ private HiveMetastore createUserCachingMetastore(String user)
...
+            return new AggregateCacheStatsMBean(CachingHiveMetastore::getConfigValuesCache);
+        }
+
+        public class AggregateCacheStatsMBean
+        {
+            private final Function<CachingHiveMetastore, Cache<?, ?>> cacheExtractor;
+
+            public AggregateCacheStatsMBean(Function<CachingHiveMetastore, Cache<?, ?>> cacheExtractor)
+            {
+                this.cacheExtractor = cacheExtractor;
```
Comment (electrum) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/SharedHiveMetastoreCache.java:334]: Check for null

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/SharedHiveMetastoreCache.java [hunk 2]
```diff
@@ -212,6 +217,201 @@ private HiveMetastore createUserCachingMetastore(String user)
...
+            }
+
+            @Managed
+            public long size()
+            {
+                long size = 0;
+                for (CachingHiveMetastore metastore : cache.asMap().values()) {
+                    size = saturatedAdd(size, cacheExtractor.apply(metastore).size());
+                }
+                return cache.size();
```
Comment (electrum) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/SharedHiveMetastoreCache.java:344]: Should return `size`

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/SharedHiveMetastoreCache.java [hunk 3]
```diff
@@ -212,6 +217,201 @@ private HiveMetastore createUserCachingMetastore(String user)
...
+
+            public AggregateCacheStatsMBean(Function<CachingHiveMetastore, Cache<?, ?>> cacheExtractor)
+            {
+                this.cacheExtractor = cacheExtractor;
+            }
+
+            @Managed
+            public long size()
+            {
+                long size = 0;
```
Comment (electrum) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/SharedHiveMetastoreCache.java:340]:
```java
return cache.asMap().values().stream()
        .map(cacheExtractor::apply)
        .mapToLong(Cache::size)
        .reduce(0, LongMath::saturatedAdd);

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java [hunk 4]
```diff
@@ -427,7 +421,10 @@ private void verifyTableNotExists(String newDatabaseName, String newTableName)
     }
 
     @Override
-    public synchronized void updateTableStatistics(HiveIdentity identity, String databaseName, String tableName, AcidTransaction transaction, Function<PartitionStatistics, PartitionStatistics> update)
+    public synchronized void updateTableStatistics(String databaseName,
```
Comment (electrum) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java:424]: Revert wrapping

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueHiveMetastore.java [hunk 5]
```diff
@@ -357,13 +356,16 @@ public PartitionStatistics getTableStatistics(HiveIdentity identity, Table table
     }
 
     @Override
-    public void updateTableStatistics(HiveIdentity identity, String databaseName, String tableName, AcidTransaction transaction, Function<PartitionStatistics, PartitionStatistics> update)
+    public void updateTableStatistics(String databaseName,
```
Comment (electrum) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueHiveMetastore.java:359]: Revert wrapping

**Review by electrum — APPROVED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/DecoratedHiveMetastoreModule.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        extends AbstractConfigurationAwareModule
+{
+    @Override
+    protected void setup(Binder binder)
+    {
+        newSetBinder(binder, HiveMetastoreDecorator.class);
+        install(new RecordingHiveMetastoreDecoratorModule());
+
+        configBinder(binder).bindConfig(CachingHiveMetastoreConfig.class);
+        binder.bind(SharedHiveMetastoreCache.class).in(javax.inject.Singleton.class);
```
[the remark this answers is not in the document]
Comment (dain) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/DecoratedHiveMetastoreModule.java:54]: Weird autocomplete error.  I didn't know that was legal

**Pull request merged**